Pool incentives grpc test

hieuvubk opened the pull request:
## Refactor some tests in pool-incentives grpc_test.go to table drive
 
  - *TestGaugeIds*
  - *TestDistrInfo*
  - *TestIncentivizedPools*

stackman27 commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -150,140 +172,123 @@ func (suite *KeeperTestSuite) TestLockableDurations() {
+				suite.NoError(err)
+				suite.Equal(tc.expectedRecordLength, len(res.IncentivizedPools))
+
+				for i := 0; i < 0; i++ {
```
Is this loop necessary?

stackman27 commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -150,140 +172,123 @@ func (suite *KeeperTestSuite) TestLockableDurations() {
+				sort.Slice(distRecords[:], func(i, j int) bool {
+					return distRecords[i].GaugeId < distRecords[j].GaugeId
+				})
+				// Create records
```
```suggestion
				// update records and ensure that non-perpetuals pot cannot get rewards. 
```

hieuvubk replied:
thanks you

stackman27 commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -150,140 +172,123 @@ func (suite *KeeperTestSuite) TestLockableDurations() {
+				}
+				tc.distRecords = append(distRecords, tc.distRecords...)
+
+				// Sort in ascending order of gaugeId
```
Why do we have put distrRecords in ascending order?

hieuvubk replied:
I believe the UpdateDistrRecords function only accepts sorted records

mattverse replied:
yeah it errors if not

stackman27 commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -35,89 +30,116 @@ func (suite *KeeperTestSuite) TestGaugeIds() {
+	} {
+		tc := tc
+		suite.Run(tc.desc, func() {
+			if tc.poolCreated {
```
is `poolCreated` = false the same as `len(weights).= 0` 

We can have a valid pool without any tokens locked right?

stackman27 commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -35,89 +30,116 @@ func (suite *KeeperTestSuite) TestGaugeIds() {
+
+					gauge, err := suite.App.IncentivesKeeper.GetGaugeByID(suite.Ctx, res.GaugeIdsWithDuration[i].GaugeId)
+					suite.NoError(err)
+					suite.Equal(0, len(gauge.Coins))
```
nit: I think the verdict was to use `suite.Require()` whenever necessary [link](https://github.com/osmosis-labs/osmosis/pull/2046#pullrequestreview-1041071475)

alexanderbez replied:
yup!!!

czarcas7ic commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -21,103 +22,124 @@ func (suite *KeeperTestSuite) TestGaugeIds() {
+			err:     true,
+		},
+		{
+			desc: "Unexisted pool",
```
```suggestion
			desc: "Nonexistent pool",
```

czarcas7ic commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -21,103 +22,124 @@ func (suite *KeeperTestSuite) TestGaugeIds() {
+				suite.Require().Error(err)
+			} else {
+				suite.Require().NoError(err)
+				suite.Require().NoError(err)
```
```suggestion
```

was this accidental duplication?

czarcas7ic commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -21,103 +22,124 @@ func (suite *KeeperTestSuite) TestGaugeIds() {
+		expectedRecordLength int
+	}{
+		{
+			desc:                 "No pool exist",
```
```suggestion
			desc:                 "No pool exists",
```

czarcas7ic commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -21,103 +22,124 @@ func (suite *KeeperTestSuite) TestGaugeIds() {
+	} {
+		tc := tc
+		suite.Run(tc.desc, func() {
+			if tc.poolCreated {
```
It appears as though we aren't even testing the poolCreated: false test right? And it seems like we are just using poolCreated as an "is an err expected" variable. If this is the case, can we keep this to similar semantics like "err","valid", or "errExpected".

hieuvubk replied:
I use that variable to test the case that no pool is created. 
Actually DistrInfo, IncentivizedPools query always returns nil error so we will check through response length and total weight. In no pool case, length and weights both be 0.

czarcas7ic replied:
I see now, thanks!

czarcas7ic commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -138,152 +160,136 @@ func (suite *KeeperTestSuite) TestLockableDurations() {
+				}
+			}
+			res, err := queryClient.IncentivizedPools(context.Background(), &types.QueryIncentivizedPoolsRequest{})
+			if !tc.expectedErr {
```
It looks like we aren't requiring errors for expectedErr = true cases right?

mattverse replied:
yeah we should add the case fpr Require Error

alexanderbez commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -21,103 +22,124 @@ func (suite *KeeperTestSuite) TestGaugeIds() {
 	// LockableDurations should be 1, 3, 7 hours from the default genesis state.
 	lockableDurations := suite.App.PoolIncentivesKeeper.GetLockableDurations(suite.Ctx)
-	suite.Equal(3, len(lockableDurations))
+	suite.Require().Equal(3, len(lockableDurations))
```
Yeeah we should be using `Require()` so the tests exit upon failure.

mattverse replied:
ahhh I see thx for the explanation!

hieuvubk commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -21,103 +22,124 @@ func (suite *KeeperTestSuite) TestGaugeIds() {
+				for i := 0; i < len(res.GaugeIdsWithDuration); i++ {
+					suite.Require().Equal(lockableDurations[i], res.GaugeIdsWithDuration[i].Duration)
+
+					gauge, err := suite.App.IncentivesKeeper.GetGaugeByID(suite.Ctx, res.GaugeIdsWithDuration[i].GaugeId)
```
This code seems unnecessary because it was tested in gauges_test. I will remove the detailed check of each gauge above

hieuvubk replied:
In the old test, there was also a detailed check of each gauge, should we remove them? Will make tests look clearer

hieuvubk commented:
I just refactored my tests, removed unnecessary loops.
About expErr variables in DistrInfo, IncentivizedPools test, functions always return nil error. So I will check through record length and total weight
Could u pls take a look @czarcas7ic @mattverse

czarcas7ic commented on x/pool-incentives/keeper/grpc_query_test.go:
```diff
@@ -170,81 +156,53 @@ func (suite *KeeperTestSuite) TestLockableDurations() {
-		expectedErr          bool
+		weights              []sdk.Int
+		perpetual		     bool
+		nonPerpetual		 bool
```
if perpetual is false that means nonPerpetual is automatically true right?

hieuvubk replied:
Ahh just optional variables to test cases with additional Perpetual or NonPerpetual gauges

czarcas7ic commented:
Thanks for the ping @hieuvubk sorry about how long it took me to get back to this. Would like one more approval and then we can merge!

ValarDragon reviewed (approved):
Nice PR, thanks for making it!

Not sure why it got stuck, sorry for that delay
